fofpetaio.c

```diff
@@ -77,13 +73,27 @@ void fof_save_particles(int num) {
     MPI_Barrier(MPI_COMM_WORLD);
     fof_write_header(&bf); 
 
+    int stype = -1;
+    int * argind = NULL;
+    int NumPIG = 0;
+
     for(i = 0; i < IOTable.used; i ++) {
         /* only process the particle blocks */
         char blockname[128];
         int ptype = IOTable.ent[i].ptype;
         BigArray array = {0};
         if(ptype < 6 && ptype >= 0) {
+            walltime_measure("/FOF/IO/Write");
             sprintf(blockname, "%d/%s", ptype, IOTable.ent[i].name);
+            if(ptype != stype) {
```

Comment:
What about creating 6 arrays before the loop.  The total length is NumPart, so they will be shifted aliases of a single buffer, with the shifting calculated by return value of fof_build_selection? It may look straighter than the a single level LRU cache..